Added network_interface config option.

pengale opened the pull request:
Allows an operator to bind kafka to listen on a specific network
interface. This addresses scenarios where one might want to restrict the
interface kafka talks on for security purposes.

@juju-solutions/bigdata

pengale committed:
Added network_interface config option.

Allows an operator to bind kafka to listen on a specific network
interface. This addresses scenarios where one might want to restrict the
interface kafka talks on for security purposes.

johnsca commented on tests/01-deploy.py:
```diff
@@ -27,6 +28,45 @@ def test_deploy(self):
+        """
+        self.d.log.debug("Binding Kafka to 0.0.0.0")
+        self.d.configure('kafka', {'network_interface': '0.0.0.0'})
+        self.d.sentry.wait_for_messages({'kafka': 'Server config changed: restarting Kafka'}, timeout=60)
```
`wait_for_messages` does not use `status-history`, so if this message is transient, it could easily be missed during the sleeping period causing intermittent failures.  I think it would be good to propose an improvement to `wait_for_messages` to use `status-history`, but in the meantime, we probably need to do this manually in the test.

pengale replied:
johnsca: the tricky thing with using status history is that we don't want to catch previous instances of this message, that have been triggered by previous tests. We don't officially know whether this test will execute first, second, or third, for example, so we don't know whether we're looking for one, two or three instances of the string.

Maybe it would make sense to just catch timeout errors here, and attempt the next part of the test; it will fail if the code hasn't been run, anyway.

kwmonroe commented on lib/charms/layer/apache_kafka.py:
```diff
@@ -135,3 +152,48 @@ def stop(self):
+            interfaces = netifaces.interfaces()
+            for interface in interfaces:
+                try:
+                    ip = netifaces.ifaddresses(interface)[2][0]['addr']
```
Instead of hard coding [2] here, netifaces suggests using `netifaces.AF_INET` to get the ipv4 entry:

https://pypi.python.org/pypi/netifaces

We'll want to handle ipv6 too, so maybe try getting the ip4 addr here, then try for ip6 (`netifaces.AF_INET6`), and then except your fate .

johnsca replied:
Yeah, I'm ending up with it binding on IPv6 in my test.  Also, @kwmonroe clearly means "accept".  ;)

pengale replied:
Yeah. Paying attention to ipv6 is probably a good idea.

Of course, now you've made work for me in the bigtop charms, too :-p

kwmonroe replied:
I actually meant `except` @johnsca.. As in, `try / except` :)

johnsca replied:
Aren't you so clever.  :p

kwmonroe commented on lib/charms/layer/apache_kafka.py:
```diff
@@ -135,3 +152,48 @@ def stop(self):
+                    network_interface))
+        else:
+            try:
+                ip = netifaces.ifaddresses(network_interface)[2][0]['addr']
```
Same as above comment, can we not hard code `[2]`?

pengale committed:
Fixes based on PR.

Now support ipv6 a little better, and don't have a failing test due to
erroneous timeouts.

kwmonroe commented:
LGTM - thanks for handling AF_INET and AF_INET6 when a CIDR is passed.  You mentioned the support isn't there to check for an INET6 address when passing an interface name.  I'm cool with making that a future feature if people ask for that.

johnsca commented:
:+1: from me
